src/coder.js

```diff
@@ -0,0 +1,166 @@
+'use strict'
+
+const varint = require('varint')
+const through = require('pull-through')
+
+const debug = require('debug')
+
+const log = debug('pull-plex:coder')
+log.err = debug('pull-plex:coder:err')
+
+const { MAX_MSG_SIZE } = require('./consts')
+
+const PULL_LENGTH = 10 * 1024
+const empty = Buffer.alloc(0)
+
+/**
+ * Creates a Through PullStream that will varint encode all
+ * messages passed through it.
+ *
+ * @returns {PullStream} A through stream that varint encodes all messages
+ */
+exports.encode = () => {
+  let pool = Buffer.alloc(PULL_LENGTH)
+  let used = 0
+
+  return through(function (msg) {
+    const oldUsed = used
+    varint.encode(msg[0] << 3 | msg[1], pool, used)
+    used += varint.encode.bytes
+    varint.encode(msg[2] ? msg[2].length : 0, pool, used)
+    used += varint.encode.bytes
+    this.queue(pool.slice(oldUsed, used)) // send header
+
+    if (PULL_LENGTH - used < 100) {
+      pool = Buffer.alloc(PULL_LENGTH)
```

Comment:
I would be bold and go for `allocUnsafe`. We are writing 100% of the memory, minus bugs.